components/base/kinematicbase/execution.go

```diff
@@ -0,0 +1,471 @@
+		defer cancelFn()
+		return multierr.Combine(errToWrap, ptgk.Base.Stop(stopCtx, nil))
+	}
+
+	startPose := spatialmath.NewZeroPose() // This is the location of the base at call time
+	if ptgk.Localizer != nil {
+		startPoseInFrame, err := ptgk.Localizer.CurrentPosition(ctx)
+		if err != nil {
+			return tryStop(err)
+		}
+		startPose = startPoseInFrame.Pose()
+	}
+
+	// Pre-process all steps into a series of velocities
+	ptgk.inputLock.Lock()
+	ptgk.currentExecutingSteps, err = ptgk.arcStepsFromInputs(inputSteps, startPose)
+	arcSteps := ptgk.currentExecutingSteps
+	ptgk.inputLock.Unlock()
+	if err != nil {
+		return tryStop(err)
+	}
+
+	for i := 0; i < len(arcSteps); i++ {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		step := ptgk.currentExecutingSteps[i]
+		ptgk.inputLock.Lock() // In the case where there's actual contention here, this could cause timing issues; how to solve?
+		ptgk.currentIdx = i
+		ptgk.currentInputs = step.arcSegment.StartConfiguration
+		ptgk.inputLock.Unlock()
+
+		timestep := time.Duration(step.durationSeconds*1000*1000) * time.Microsecond
+
+		ptgk.logger.Debugf("step, i %d", i)
+		ptgk.logger.Debug(
+			step.linVelMMps,
+			step.angVelDegps,
+			step.durationSeconds,
+			step.arcSegment,
+			spatialmath.PoseToProtobuf(step.arcSegment.StartPosition),
+		)
+
+		ptgk.logger.CDebugf(ctx,
+			"setting velocity to linear %v angular %v and running velocity step for %s",
+			step.linVelMMps,
+			step.angVelDegps,
+			timestep,
+		)
+
+		err := ptgk.Base.SetVelocity(
+			ctx,
+			step.linVelMMps,
+			step.angVelDegps,
+			nil,
+		)
+		if err != nil {
+			return tryStop(err)
+		}
+		arcStartTime := time.Now()
+		// Now we are moving. We need to do several things simultaneously:
+		// - move until we think we have finished the arc, then move on to the next step
+		// - update our CurrentInputs tracking where we are through the arc
+		// - Check where we are relative to where we think we are, and tweak velocities accordingly
+		for timeElapsed := inputUpdateStepSeconds; timeElapsed <= step.durationSeconds; timeElapsed += inputUpdateStepSeconds {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			// Account for 1) timeElapsed being inputUpdateStepSeconds ahead of actual elapsed time, and the fact that the loop takes
+			// nonzero time to run especially when using the localizer.
+			actualTimeElapsed := time.Since(arcStartTime)
+			remainingTimeStep := time.Duration(1000*1000*timeElapsed)*time.Microsecond - actualTimeElapsed
+
+			if remainingTimeStep > 0 {
+				utils.SelectContextOrWait(ctx, remainingTimeStep)
+				if ctx.Err() != nil {
+					return tryStop(ctx.Err())
+				}
+			}
+			distIncVel := step.linVelMMps.Y
+			if distIncVel == 0 {
+				distIncVel = step.angVelDegps.Z
+			}
+			currentInputs := []referenceframe.Input{
+				step.arcSegment.StartConfiguration[0],
+				step.arcSegment.StartConfiguration[1],
+				{step.arcSegment.StartConfiguration[2].Value + math.Abs(distIncVel)*timeElapsed},
```

Comment:
This is how distance is defined for PTGs; the linear distance, or if in a region where linear velocity is zero, the angular distance in degrees. Without this it would not be possible to track intermediate points during rotations in place.

Reply:
I understand, but I think thats a little fishy since youre adding things that have different units, and might be related to why spins dont work well.  Just something to keep in the back of our heads as we try to solve that problem